ubuntu_image/tests/test_builder.py

```diff
@@ -826,6 +826,121 @@ def test_make_disk_with_parts(self):
             self.assertEqual(partitions[1], ('beta', 8192))
             self.assertEqual(partitions[2], ('writable', 10240))
 
+    def test_make_disk_with_parts_out_of_order(self):
+        # Structures get partition numbers matching their appearance in the
+        # gadget.yaml, even if these are out of order with respect to their
+        # disk offsets.  LP: #1642999
+        with ExitStack() as resources:
+            workdir = resources.enter_context(TemporaryDirectory())
+            unpackdir = resources.enter_context(TemporaryDirectory())
+            # Fast forward a state machine to the method under test.
+            args = SimpleNamespace(
+                cloud_init=None,
+                output=None,
+                unpackdir=unpackdir,
+                workdir=workdir,
+                )
+            # Jump right to the method under test.
+            state = resources.enter_context(XXXModelAssertionBuilder(args))
+            state._next.pop()
+            state._next.append(state.make_disk)
+            # Set up expected state.
+            state.unpackdir = unpackdir
+            state.images = os.path.join(workdir, '.images')
+            state.disk_img = os.path.join(workdir, 'disk.img')
+            state.image_size = MiB(10)
+            os.makedirs(state.images)
+            # Craft a gadget schema.
+            part0 = SimpleNamespace(
+                name='alpha',
+                type='aa',
+                size=MiB(1),
+                offset=MiB(2),
+                offset_write=100,
+                )
+            part1 = SimpleNamespace(
+                name='beta',
+                type='bb',
+                size=MiB(1),
+                offset=MiB(4),
+                offset_write=200,
+                )
+            part2 = SimpleNamespace(
+                name='gamma',
+                type='mbr',
```

Comment:
I haven't yet done a wholesale change of the test suite to use `role` yet.  This is mostly a copy-n-paste of an existing test.  [LP: #1643086](https://bugs.launchpad.net/ubuntu-image/+bug/1643086)